### PR TITLE
[operator-trivy] Distroless

### PR DESCRIPTION
**Pull request by zuzzas**
## Description

Changed all images to use distroless image by default.

## Why do we need it, and what problem does it solve?

Using images with nothing inside them except statically linked program improves security.

## What is the expected result?

operator-trivy runs without errors and all Jobs created by it succeed and create appropriate CRs.

```
root@klimentyev-master-0:~# kubectl -n d8-operator-trivy get pod
NAME                                        READY   STATUS      RESTARTS   AGE
operator-66c658fbb5-dg4fb                   2/2     Running     0          39m

root@klimentyev-master-0:~# kubectl -n d8-operator-trivy get vulnerabilityreports.aquasecurity.github.io -A
NAMESPACE              NAME                                                       REPOSITORY          TAG   SCANNER   AGE
d8-cni-simple-bridge   daemonset-simple-bridge-simple-bridge                      sys/deckhouse-oss         Trivy     65s
d8-descheduler         replicaset-556f7b9849                                      sys/deckhouse-oss         Trivy     18s
d8-monitoring          statefulset-prometheus-longterm-config-reloader            sys/deckhouse-oss         Trivy     64s
d8-monitoring          statefulset-prometheus-longterm-init-config-reloader       sys/deckhouse-oss         Trivy     64s
d8-monitoring          statefulset-prometheus-longterm-kube-rbac-proxy            sys/deckhouse-oss         Trivy     64s
d8-monitoring          statefulset-prometheus-longterm-prometheus                 sys/deckhouse-oss         Trivy     64s
d8-operator-trivy      replicaset-operator-66c658fbb5-kube-rbac-proxy             sys/deckhouse-oss         Trivy     14s

root@klimentyev-master-0:~# kubectl -n d8-operator-trivy get rbacassessmentreports.aquasecurity.github.io -A
NAMESPACE                    NAME                                                            SCANNER   AGE
argocd                       role-argocd-application-controller                              Trivy     2d23h
argocd                       role-argocd-applicationset-controller                           Trivy     2d23h
argocd                       role-argocd-dex-server                                          Trivy     2d23h
argocd                       role-argocd-notifications-controller                            Trivy     2d23h
argocd                       role-argocd-server                                              Trivy     2d23h
d8-admission-policy-engine   role-access-to-admission-policy-engine                          Trivy     2d23h
d8-admission-policy-engine   role-gatekeeper-manager                                         Trivy     2d23h

kubectl -n d8-operator-trivy get infraassessmentreports.aquasecurity.github.io -A
NAMESPACE     NAME                                              SCANNER   AGE
kube-system   pod-etcd-klimentyev-master-0                      Trivy     2d23h
kube-system   pod-kube-apiserver-klimentyev-master-0            Trivy     2d23h
kube-system   pod-kube-controller-manager-klimentyev-master-0   Trivy     2d23h
kube-system   pod-kube-scheduler-klimentyev-master-0            Trivy     2d23h
```

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: operator-trivy
type: feature
summary: All images are now based on distroless image
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've started the workflow for label 'e2e/run/gcp'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've started the workflow for label 'e2e/run/aws'...

File: ee/modules/500-operator-trivy/images/node-collector/Dockerfile [hunk 1]
```diff
@@ -1,17 +1,20 @@
...
-ARG BASE_GOLANG_19_ALPINE
+ARG BASE_DISTROLESS
+ARG BASE_GOLANG_20_ALPINE
 
-FROM $BASE_GOLANG_19_ALPINE AS build
+FROM $BASE_GOLANG_20_ALPINE AS build
+ARG SOURCE_REPO
+ENV SOURCE_REPO=${SOURCE_REPO}
+ARG GOPROXY
+ENV GOPROXY=${GOPROXY}
```
Comment (RomanenkoDenys) [inline ee/modules/500-operator-trivy/images/node-collector/Dockerfile:8]:
```suggestion
ARG GOPROXY
ARG SOURCE_REPO

ENV GOPROXY=${GOPROXY} \
    SOURCE_REPO=${SOURCE_REPO} \
    CGO_ENABLED=0 \
    GOOS=linux \
    GOARCH=amd64
```

File: ee/modules/500-operator-trivy/images/node-collector/Dockerfile [hunk 2]
```diff
@@ -1,17 +1,20 @@
...
 WORKDIR /src
-RUN apk add --update --no-cache patch
-RUN wget https://github.com/aquasecurity/k8s-node-collector/archive/refs/tags/v0.0.6.tar.gz -O - | tar -xz --strip-components=1
+RUN apk add --update --no-cache patch git
+RUN git clone --depth 1 --branch v0.0.6 ${SOURCE_REPO}/aquasecurity/k8s-node-collector.git .
 
 COPY patches/001-change-node-collector-config.patch /src
 RUN patch -p1 < 001-change-node-collector-config.patch
 
 RUN CGO_ENABLED=0 GOOS=linux go build -ldflags '-s -w -extldflags "-static"' -o node-collector ./cmd/node-collector/main.go
```
Comment (RomanenkoDenys) [inline ee/modules/500-operator-trivy/images/node-collector/Dockerfile:16]:
```suggestion
RUN go build -ldflags '-s -w -extldflags "-static"' -o node-collector ./cmd/node-collector/main.go
```

File: ee/modules/500-operator-trivy/images/operator/Dockerfile [hunk 3]
```diff
@@ -1,10 +1,14 @@
...
-ARG BASE_GOLANG_19_ALPINE
+ARG BASE_DISTROLESS
+ARG BASE_GOLANG_20_ALPINE
 
-FROM $BASE_GOLANG_19_ALPINE AS build
+FROM $BASE_GOLANG_20_ALPINE AS build
+ARG SOURCE_REPO
+ENV SOURCE_REPO=${SOURCE_REPO}
+ARG GOPROXY
+ENV GOPROXY=${GOPROXY}
```
Comment (RomanenkoDenys) [inline ee/modules/500-operator-trivy/images/operator/Dockerfile:8]:
```suggestion
ARG GOPROXY
ARG SOURCE_REPO

ENV GOPROXY=${GOPROXY} \
    SOURCE_REPO=${SOURCE_REPO} \
    CGO_ENABLED=0 \
    GOOS=linux \
    GOARCH=amd64
```

File: ee/modules/500-operator-trivy/images/operator/Dockerfile [hunk 4]
```diff
@@ -14,7 +18,9 @@ RUN patch -p1 < 001-add-registry-secret-as-dockerconfigjson.patch && \
 
 RUN CGO_ENABLED=0 GOOS=linux go build -ldflags '-s -w -extldflags "-static"' -o operator-trivy ./cmd/trivy-operator/main.go
```
Comment (RomanenkoDenys) [inline ee/modules/500-operator-trivy/images/operator/Dockerfile:19]:
```suggestion
RUN go build -ldflags '-s -w -extldflags "-static"' -o operator-trivy ./cmd/trivy-operator/main.go
```

File: ee/modules/500-operator-trivy/images/trivy/Dockerfile [hunk 5]
```diff
@@ -1,19 +1,23 @@
...
 ARG SOURCE_REPO
 ENV SOURCE_REPO=${SOURCE_REPO}
+ARG GOPROXY
+ENV GOPROXY=${GOPROXY}
 
 WORKDIR /src
 RUN git clone --depth 1 --branch ${TRIVY_DB_VERSION} ${SOURCE_REPO}/aquasecurity/trivy-db.git && \
     git clone --depth 1 --branch ${TRIVY_VERSION} ${SOURCE_REPO}/aquasecurity/trivy.git && \
     cd trivy && \
     CGO_ENABLED=0 GOOS=linux go build -ldflags '-s -w -extldflags "-static"' -o trivy ./cmd/trivy/main.go
```
Comment (RomanenkoDenys) [inline ee/modules/500-operator-trivy/images/trivy/Dockerfile:16]:
```suggestion
    go build -ldflags '-s -w -extldflags "-static"' -o trivy ./cmd/trivy/main.go
```

**Review by RomanenkoDenys — COMMENTED** [review]
What do you think about different format of variables ?

**Review by name212 — COMMENTED** [review]
confirm with Denis

File: ee/modules/500-operator-trivy/images/trivy/Dockerfile (hunk 5 above)
Comment (zuzzas) [reply ee/modules/500-operator-trivy/images/trivy/Dockerfile]: Fixed.

File: ee/modules/500-operator-trivy/images/trivy/Dockerfile [hunk 6]
```diff
@@ -1,19 +1,23 @@
...
+ARG BASE_DISTROLESS
 ARG BASE_GOLANG_19_BULLSEYE
 
 FROM $BASE_GOLANG_19_BULLSEYE AS build
 ARG TRIVY_VERSION=v0.44.0-flant
 ARG TRIVY_DB_VERSION=flant-latest
 ARG SOURCE_REPO
 ENV SOURCE_REPO=${SOURCE_REPO}
+ARG GOPROXY
+ENV GOPROXY=${GOPROXY}
```
[the remark this answers is not in the document]
Comment (zuzzas) [reply ee/modules/500-operator-trivy/images/trivy/Dockerfile]: Fixed.

File: ee/modules/500-operator-trivy/images/operator/Dockerfile (hunk 4 above)
Comment (zuzzas) [reply ee/modules/500-operator-trivy/images/operator/Dockerfile]: Fixed.

File: ee/modules/500-operator-trivy/images/operator/Dockerfile (hunk 3 above)
Comment (zuzzas) [reply ee/modules/500-operator-trivy/images/operator/Dockerfile]: Fixed.

File: ee/modules/500-operator-trivy/images/node-collector/Dockerfile (hunk 2 above)
Comment (zuzzas) [reply ee/modules/500-operator-trivy/images/node-collector/Dockerfile]: Fixed.

File: ee/modules/500-operator-trivy/images/node-collector/Dockerfile (hunk 1 above)
Comment (zuzzas) [reply ee/modules/500-operator-trivy/images/node-collector/Dockerfile]: Fixed.

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've started the workflow for label 'e2e/run/gcp'...

**Review by RomanenkoDenys — DISMISSED** [review]
LGTM

File: ee/modules/500-operator-trivy/images/node-collector/Dockerfile [hunk 7]
```diff
@@ -1,17 +1,24 @@
...
-RUN apk add --update --no-cache patch
-RUN wget https://github.com/aquasecurity/k8s-node-collector/archive/refs/tags/v0.0.6.tar.gz -O - | tar -xz --strip-components=1
+RUN apk add --update --no-cache patch git
+RUN git clone --depth 1 --branch v0.0.6 ${SOURCE_REPO}/aquasecurity/k8s-node-collector.git .
 
 COPY patches/001-change-node-collector-config.patch /src
 RUN patch -p1 < 001-change-node-collector-config.patch
 
-RUN CGO_ENABLED=0 GOOS=linux go build -ldflags '-s -w -extldflags "-static"' -o node-collector ./cmd/node-collector/main.go
+RUN go build -ldflags '-s -w -extldflags "-static"' -o node-collector ./cmd/node-collector/main.go
```
Comment (aleksey-su) [inline ee/modules/500-operator-trivy/images/node-collector/Dockerfile:20]: The rights for the deckhouse user are not set

Comment (zuzzas) [reply ee/modules/500-operator-trivy/images/node-collector/Dockerfile:20]: It does not work without root rights. I'd like to merge this PR and go and check all still-left-root-permissions after all distroless stuff is merged.

**Review by aleksey-su — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've started the workflow for label 'e2e/run/aws'...

**Review by name212 — APPROVED** [review]

**Pull request merged**